sda/router.go

```diff
@@ -0,0 +1,543 @@
+	// start processing messages
+	go func() {
+		t.networkLock.Lock()
+		t.quitProcessMsg = make(chan bool)
+		if t.processMessagesStarted {
+			// we are already listening
+			t.networkLock.Unlock()
+			return
+		}
+		t.networkLock.Unlock()
+
+		t.processMessages()
+	}()
+	// listen
+	go func() {
+		err := t.listen()
+		if err != nil {
+			log.Fatal("Error listening on", t.workingAddress, ":", err)
+		}
+	}()
+	<-t.closing
+}
+
+func (t *TCPRouter) listen() error {
+	// open port
+	log.Lvl3(t.serverIdentity.First(), "starts to listen")
+	fn := func(c network.SecureConn) {
+		log.Lvl3(t.workingAddress, "Accepted Connection from", c.Remote())
+		// register the connection once we know it's ok
+		t.registerConnection(c)
+		t.handleConn(c)
+	}
+	err := t.host.Listen(fn)
+	if err != nil {
+		log.Fatal("Error listening on", t.workingAddress, ":", err)
+		return err
+	}
+	return nil
+}
+
+// processMessages is receiving all the messages coming from the network and
+// dispatch it to the Dispatcher.
+func (t *TCPRouter) processMessages() {
+	t.networkLock.Lock()
+	t.processMessagesStarted = true
+	log.Lvl5(t.workingAddress, "Starting Process Messages")
+	t.networkLock.Unlock()
+	for {
+		var data network.Packet
+		select {
+		case <-t.quitProcessMsg:
+			log.Lvl5(t.workingAddress, "Quitting ProcessMessages")
+			t.networkLock.Lock()
+			t.processMessagesStarted = false
+			t.networkLock.Unlock()
+			return
+		case data = <-t.networkChan:
+		}
+		log.Lvl4(t.workingAddress, "Message Received from", data.From, data.MsgType)
+		switch data.MsgType {
+		case network.ErrorType:
+			log.Lvl3("Error from the network")
+		default:
+			// The dispatcher will call the appropriate processors for the
+			// message
+			if err := t.Dispatch(&data); err != nil {
+				log.Lvl3("Unknown message received:", data, err)
+			}
+		}
+	}
+}
+
+// Connect takes an entity where to connect to
+func (t *TCPRouter) Connect(id *network.ServerIdentity) (network.SecureConn, error) {
+	var err error
+	var c network.SecureConn
+	// try to open connection
+	c, err = t.host.Open(id)
+	if err != nil {
+		return nil, err
+	}
+	log.Lvl3("Host", t.workingAddress, "connected to", c.Remote())
+	t.registerConnection(c)
+	go t.handleConn(c)
+	return c, nil
+}
+
+// Close shuts down all network connections and closes the listener.
+func (t *TCPRouter) Close() error {
+	t.closingMut.Lock()
+	defer t.closingMut.Unlock()
+	if t.isClosing {
+		return errors.New("Already closing")
+	}
+	log.Lvl4(t.serverIdentity.First(), "Starts closing")
+	t.isClosing = true
+	// stop the Run
+	t.closing <- true
+	t.networkLock.Lock()
+	if t.processMessagesStarted {
+		// Tell ProcessMessages to quit
+		close(t.quitProcessMsg)
+	}
+	t.networkLock.Unlock()
+	// The tcp host is supposed to take care of the connection for us
+	if err := t.host.Close(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// closeConnection closes a connection and removes it from the connections-map
+// The t.networkLock must be taken.
```

Comment:
OK right I've added a new lock then only for the processing of the message. Now networkLock is only used for the connections.